### PR TITLE
Support batch mutiple steps into a single tx

### PR DESCRIPTION
**Pull request by tolak**
## The implementation

image

In this PR, we introduced `MultiStep` which represents either a single step or a batch of steps. We merge steps in which execution transactions can happen on the same source chain during task initialization, even before applying nonce for each step. Of course, now we apply nonce for each MultiStep rather than each Step.
This can be done because we have support real-time settlement on our handler contract where we will update the input spend value of the next step while we execute each step.

image

Now we have several `Action`s (previous executor), each of them representing one or more on-chain operations. Each object in `StepJson` will specify the action ID so we can look up the action during we convert it to `Step` object. Each `Action` can produce a bunch of calls according to the `Step` data, so by executing a `MultiStep` we just need to combine them together and send it to the blockchain through Handler method `batchCall`.

## Others

To make the codebase simple and easy to maintain, the old `index` directory was removed, and the `Executor` now moves to `actions` directory and separated in each chain folder.

We also removed `examples` directory, because we already have `awesome-phat-contract` repo that contains well-maintained examples.

### CODE REVIEW

File: contracts/index_executor/src/task.rs [hunk 1]
```diff
@@ -257,6 +321,90 @@ impl Task {
...
+                    } else {
+                        Some(step_source_chain.handler_contract.clone())
+                    }
+                }
+            };
+        }
+        Ok(())
+    }
+
+    pub fn merge_step(&mut self, context: &Context) -> Result<(), &'static str> {
```
Comment (h4x3rotab) [inline contracts/index_executor/src/task.rs:361]: Haven't fully review it. Would be nice if you can add some comments to describe how it works.

File: contracts/index_executor/src/task.rs [hunk 2]
```diff
@@ -257,6 +321,90 @@ impl Task {
...
     }
 
+    pub fn apply_recipient(&mut self, context: &Context) -> Result<(), &'static str> {
+        let step_count = self.steps.len();
+        for (index, step) in self.steps.iter_mut().enumerate() {
+            let step_source_chain = &step.source_chain(context).ok_or("MissingChain")?;
+            let step_dest_chain = &step.dest_chain(context).ok_or("MissingChain")?;
+            let worker_account_info = context.get_account(self.worker).ok_or("WorkerNotFound")?;
+
+            // For sure last step we should put real recipient, or else the recipient could be either
```
Comment (h4x3rotab) [inline contracts/index_executor/src/task.rs:331]: According to our previous discussion, all the recipient should be the worker account, unless it uses the handler to temporarily control the assets for swap and bridge, right?

File: contracts/index_executor/src/task.rs [hunk 3]
```diff
@@ -272,221 +420,396 @@ impl Task {
...
+        if self.claim_nonce.is_none() || !self.has_claimed(context)? {
+            let claim_nonce = self.get_nonce(context, &self.source)?;
+            nonce_map.insert(self.source.clone(), &(claim_nonce + 1));
+            self.claim_nonce = Some(claim_nonce);
+        }
+
+        // Apply nonce for each step
+        for index in start_index as usize..self.merged_steps.len() {
+            let nonce: u64 =
+                match nonce_map.get(&self.merged_steps[index].as_single_step().source_chain) {
```
Comment (h4x3rotab) [inline contracts/index_executor/src/task.rs:435]: Nit: Let's make the key a variable so you don't need to write `merged_steps[index].as_single_step().source_chain` multiple times.

**Review by h4x3rotab — COMMENTED** [review]
Let me submit the first round review. I haven't finished "tasks.rs" yet. WIll add more comments later.

File: contracts/index_executor/src/task.rs [hunk 4]
```diff
@@ -64,24 +85,50 @@ impl Default for Task {
...
 
-        if client.read::<Task>(&self.id)?.is_some() {
-            // Task already saved, return
-            return Ok(());
-        }
+        pink_extension::debug!(
+            "Trying to lookup storage for task {:?} before initializing.",
+            hex::encode(self.id),
+        );
+        // if client.read_storage::<Task>(&self.id)?.is_some() {
```
Comment (h4x3rotab) [inline contracts/index_executor/src/task.rs:121]: Nit: remove comments if not used

File: contracts/index_executor/src/task.rs [hunk 5]
```diff
@@ -272,221 +420,396 @@ impl Task {
...
 
-    fn settle(&mut self, context: &Context) -> Result<u128, &'static str> {
-        if self.execute_index < 1 {
-            return Err("InvalidExecuteIndex");
-        }
+    fn claim(&mut self, context: &Context) -> Result<Vec<u8>, &'static str> {
+        let chain = context
+            .registry
+            .get_chain(&self.source)
+            .map(Ok)
```
Comment (h4x3rotab) [inline contracts/index_executor/src/task.rs:472]: Nit: Do you want `get_chain().ok_or("MissingChain")?`

File: contracts/index_executor/src/task.rs [hunk 6]
```diff
@@ -272,221 +420,396 @@ impl Task {
...
+            worker.address(),
+            Options::default(),
+        ))
+        .map_err(|e| {
+            pink_extension::error!(
+                "claimAndBatchCall: failed to estimate gas cost with error {:?}",
+                &e
+            );
+            "GasEstimateFailed"
+        })?;
```
Comment (h4x3rotab) [inline contracts/index_executor/src/task.rs:545]: Nit: It's good to use Kevin's`ResultExt` trait like this:

```rust
use pink::ResultExt;

let gas = resolve_ready(...)
    .log_err("claimAndBatchCall: failed to estimate gas cost with error")
    .or(Err("GasEstimateFailed"))?;
```

File: scripts/src/HandlerABI.json
[the remark this answers is not in the document]
Comment (tolak) [reply scripts/src/HandlerABI.json:1]: removed

File: contracts/index_executor/src/lib.rs [hunk 7]
```diff
@@ -312,14 +358,40 @@ mod index_executor {
...
         #[ink(message)]
-        pub fn get_running_task(&self, _task_id: TaskId) -> Result<Option<Task>> {
-            // TODO: read from remote storage
-            Ok(None)
+        pub fn get_running_task(&self, id: TaskId) -> Result<Option<Task>> {
+            let config = self.ensure_configured()?;
+            let client = StorageClient::new(config.db_url.clone(), config.db_token.clone());
+            Ok(client
+                .read::<Task>(&id)
+                .map_err(|_| Error::FailedToReadStorage)?
```
[the remark this answers is not in the document]
Comment (tolak) [reply contracts/index_executor/src/lib.rs:393]: Sometimes we do want to see what the error is when debug

File: contracts/index_executor/src/lib.rs [hunk 8]
```diff
@@ -312,14 +358,40 @@ mod index_executor {
...
         #[ink(message)]
         pub fn get_all_running_tasks(&self) -> Result<Vec<Task>> {
-            // TODO: read from remote storage
-            Ok(vec![])
+            let config = self.ensure_configured()?;
+            let client = StorageClient::new(config.db_url.clone(), config.db_token.clone());
+
+            let mut tasks = Vec::new();
+            let Some((ids, _)) = client
+                .read::<Vec<[u8; 32]>>(b"pending_tasks")
```
[the remark this answers is not in the document]
Comment (tolak) [reply contracts/index_executor/src/lib.rs:366]: No limit to that, it is possible to timeout in query, but we can handle it

File: contracts/index_executor/src/lib.rs [hunk 9]
```diff
@@ -119,6 +128,13 @@ mod index_executor {
             Ok(())
         }
 
+        /// Debug only, remove before release
```
[the remark this answers is not in the document]
Comment (tolak) [reply contracts/index_executor/src/lib.rs]: Done

File: contracts/index_executor/src/lib.rs [hunk 10]
```diff
@@ -243,6 +262,34 @@ mod index_executor {
             Ok(())
         }
 
+        #[ink(message)]
+        pub fn worker_drop_task(&self, worker: [u8; 32], chain: String, id: TaskId) -> Result<()> {
```
[the remark this answers is not in the document]
Comment (tolak) [reply contracts/index_executor/src/lib.rs:266]: yes, and will refund the asset user deposit when create task

File: contracts/index_executor/src/gov.rs [hunk 11]
```diff
@@ -12,6 +13,45 @@ use pink_web3::{
...
 impl WorkerGov {
+    pub fn drop_task(
+        worker_key: [u8; 32],
+        endpoint: String,
+        handler: Address,
+        id: TaskId,
+    ) -> Result<Vec<u8>, &'static str> {
+        let transport = Eth::new(PinkHttp::new(endpoint));
+        let handler_contract =
+            Contract::from_json(transport, handler, include_bytes!("./abi/handler.json"))
```
[the remark this answers is not in the document]
Comment (tolak) [reply contracts/index_executor/src/gov.rs]: Done

File: contracts/index_executor/src/gov.rs [hunk 12]
```diff
@@ -12,6 +13,45 @@ use pink_web3::{
...
+
+        let gas = resolve_ready(handler_contract.estimate_gas(
+            "drop",
+            id,
+            worker.address(),
+            Options::default(),
+        ))
+        .or(Err("GasEstimateFailed"))?;
+
+        // Submit the `approve` transaction
```
[the remark this answers is not in the document]
Comment (tolak) [reply contracts/index_executor/src/gov.rs]: Fixed

File: contracts/index_executor/src/actions/acala/asset.rs [hunk 13]
```diff
@@ -158,18 +135,28 @@ impl AcalaAssetMap {
     pub fn get_asset_attrs(
         location: &MultiLocation,
     ) -> Option<(TokenSymbol, TokenType, Option<ForeignAssetId>)> {
-        let tokens = AcalaAssetMap::get_map();
-        let token = tokens.iter().find(|s| s.0 == location.clone());
+        let tokens = AcalaAssets::get_map();
```
[the remark this answers is not in the document]
Comment (tolak) [reply contracts/index_executor/src/actions/acala/asset.rs:141]: Changed to BTreeMap

File: contracts/index_executor/src/actions/ethereum/mod.rs [hunk 14]
```diff
@@ -0,0 +1,21 @@
+use crate::actions::base::uniswapv2;
+use alloc::{boxed::Box, string::String, vec, vec::Vec};
+pub type EthereumUniswapV2 = uniswapv2::UniswapV2;
+
+use crate::call::CallBuilder;
+use crate::chain::Chain;
+use crate::utils::ToArray;
+
+pub fn create_actions(chain: &Chain) -> Vec<(String, Box<dyn CallBuilder>)> {
+    let uniswapv2_router: [u8; 20] = hex::decode("7a250d5630B4cF539739dF2C5dAcb4c659F2488D")
```
[the remark this answers is not in the document]
Comment (tolak) [reply contracts/index_executor/src/actions/ethereum/mod.rs]: Done

File: contracts/index_executor/src/call.rs [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+pub enum CallParams {
+    Evm(EvmCall),
+    Sub(SubCall),
+}
+
+#[derive(Clone, Decode, Encode, Eq, PartialEq, Ord, PartialOrd, Debug)]
+#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+pub struct Call {
+    pub params: CallParams,
+    // The call index that whose result will be the input of call
```
[the remark this answers is not in the document]
Comment (tolak) [reply contracts/index_executor/src/call.rs:55]: We don't assume all call has output, only if it need settle, it has output and can be the input of another call

File: contracts/index_executor/src/call.rs [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+    pub input_call: Option<u8>,
+    // Current call index
+    pub call_index: Option<u8>,
```
[the remark this answers is not in the document]
Comment (tolak) [reply contracts/index_executor/src/call.rs:58]: When initiate it inside the action

File: contracts/index_executor/src/call.rs [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+impl Tokenizable for Call {
+    fn from_token(_token: Token) -> Result<Self, PinkError> {
+        Err(PinkError::InterfaceUnsupported)
+    }
+
+    fn into_token(self) -> Token {
```
[the remark this answers is not in the document]
Comment (tolak) [reply contracts/index_executor/src/call.rs:66]: Done

File: contracts/index_executor/src/chain.rs [hunk 18]
```diff
@@ -162,14 +171,16 @@ impl BalanceFetcher for Chain {
                             .map_err(|_| "InvalidMultilocation")?;
                     match self.foreign_asset {
                         Some(ForeignAssetModule::PalletAsset) => {
-                            let asset_id = Location2Assetid::new()
-                                .get_assetid(self.name.clone(), &asset_location)
-                                .ok_or("AssetNotRecognized")?;
+                            let asset_id = get_assetid_by_location(&self.name, &asset_location);
+                            if asset_id.is_empty() {
```
[the remark this answers is not in the document]
Comment (tolak) [reply contracts/index_executor/src/chain.rs:175]: It's Vec<u8>, not Option

File: contracts/index_executor/src/task_deposit.rs [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+use pink_web3::{
+    contract::{tokens::Detokenize, Error as PinkError},
+    ethabi::Token,
+    types::{Address, U256},
+};
+use scale::{Decode, Encode};
+use xcm::v3::AssetId as XcmAssetId;
+
+#[derive(Debug)]
+pub struct EvmDepositData {
```
[the remark this answers is not in the document]
Comment (tolak) [reply contracts/index_executor/src/task_deposit.rs:13]: Will follow the rename when we rename the handler and others

File: contracts/index_executor/src/task_deposit.rs [hunk 20]
```diff
@@ -0,0 +1,142 @@
...
+    recipient: Vec<u8>,
+    task: String,
+}
+
+impl Detokenize for EvmDepositData {
+    fn from_tokens(tokens: Vec<Token>) -> Result<Self, PinkError>
+    where
+        Self: Sized,
+    {
+        if tokens.len() == 1 {
```
[the remark this answers is not in the document]
Comment (tolak) [reply contracts/index_executor/src/task_deposit.rs]: Done

File: contracts/index_executor/src/task_fetcher.rs [hunk 21]
```diff
@@ -0,0 +1,147 @@
...
+        .map_err(|_| "ConstructContractFailed")?;
+
+        let worker_address: Address = worker.account20.into();
+        pink_extension::debug!(
+            "Lookup actived task for worker {:?} on {:?}",
+            &hex::encode(worker_address),
+            &chain.name
+        );
+
+        let task_id: [u8; 32] = resolve_ready(handler.query(
```
[the remark this answers is not in the document]
Comment (tolak) [reply contracts/index_executor/src/task_fetcher.rs:62]: yes

File: contracts/index_executor/src/task.rs (hunk 1 above)
Comment (tolak) [reply contracts/index_executor/src/task.rs:347]: TBH, I don't see there are a lot of improvements except less if .. else .. patterns. We are moving merge to SDK BTW

File: contracts/index_executor/src/task.rs (hunk 2 above)
Comment (tolak) [reply contracts/index_executor/src/task.rs:317]: yes, considering this need refactor on the smart contract side (support batch call along with asset transfer from worker to handler), I will make a separate PR

File: contracts/index_executor/src/task.rs (hunk 3 above)
Comment (tolak) [reply contracts/index_executor/src/task.rs]: Done

File: contracts/index_executor/src/task.rs [hunk 22]
```diff
@@ -272,221 +420,396 @@ impl Task {
         context: &Context,
         _client: &StorageClient,
     ) -> Result<(), &'static str> {
-        // Only in last step the recipient with be set as real recipient on destchain,
-        // or else it will be the worker account under the hood
         let mut nonce_map: Mapping<String, u64> = Mapping::default();
```
[the remark this answers is not in the document]
Comment (tolak) [reply contracts/index_executor/src/task.rs]: Done

File: contracts/index_executor/src/task.rs [hunk 23]
```diff
@@ -272,221 +420,396 @@ impl Task {
...
-                            // ChainType::Unknown => panic!("chain not supported!"),
-                        })
-                    };
-                }
-                _ => {}
-            }
-        }
-
-        Ok(())
+    fn get_nonce(&self, context: &Context, chain: &String) -> Result<u64, &'static str> {
```
[the remark this answers is not in the document]
Comment (tolak) [reply contracts/index_executor/src/task.rs:433]: It reads `worker` actually, and following methods read more

File: contracts/index_executor/src/task.rs (hunk 4 above)
Comment (tolak) [reply contracts/index_executor/src/task.rs]: Removed

File: contracts/index_executor/src/task.rs (hunk 6 above)
Comment (tolak) [reply contracts/index_executor/src/task.rs]: Done

**Review by kingsleydon — APPROVED** [review]

**Pull request merged**